File: 프로그래머스/2/169199. 리코쳇 로봇/리코쳇 로봇.py
from collections import deque
import sys

def solution(board):
    n, m = len(board), len(board[0])
    robot, goal = None, None

    # G, R 위치 찾기
    for i in range(n):
        for j in range(m):
            if board[i][j] == "R":
                robot = (i, j)
            if board[i][j] == "G":
                goal = (i, j)
    
    # visited 초기화
    visited = [[sys.maxsize for _ in range(m)] for _ in range(n)]
    visited[robot[0]][robot[1]] = 0
    
    
    def find_next(x, y, dx, dy):
        # dx: 0, dy: 1, pos: 현재 위치
        while True:
            nx, ny = x + dx, y + dy
            if 0 <= nx < n and 0 <= ny < m and board[nx][ny] != "D":
                x, y = nx, ny
            else:
                break
        
        return x, y
        
            
    def bfs(start):        
        dx = [0, 0, 1, -1]
        dy = [1, -1, 0, 0]
        
        queue = deque()
        x, y = start
        queue.append((x, y, 0))
        
        while queue:
            x, y, m = queue.popleft()
            
            if (x, y) == goal:
                return m
            
            for i in range(4):
                nx, ny = find_next(x, y, dx[i], dy[i])
                if visited[nx][ny] > m + 1:
                    queue.append((nx, ny, m + 1))
                    visited[nx][ny] = m + 1
                    
        return -1
    
    answer = bfs(robot)
    return answer

File: 프로그래머스/2/169199. 리코쳇 로봇/리코쳇 로봇.java
import java.util.*;

class Solution {
    public int solution(String[] board) {        
        // R, G 찾기
        int[] start = new int[3];
        int[] end = new int[3];
        
        for(int i = 0; i < board.length; i++){
            for(int j = 0; j < board[i].length(); j++) {
                if(board[i].charAt(j) == 'R'){
                    start[0] = i;
                    start[1] = j;
                    start[2] = 0;
                }
                if(board[i].charAt(j) == 'G'){
                    end[0] = i;
                    end[1] = j;
                    end[2] = 0;
                }
            }
        }
        
        int answer = bfs(start, end, board);
        return answer;
    }
    
    public int bfs(int[] start, int[] end, String[] board){
        int[] dx = {0, 0, -1, 1};
        int[] dy = {-1, 1, 0, 0};
        boolean[][] visited = new boolean[board.length][board[0].length()];
        Queue<int[]> queue = new LinkedList<>();
        int count = Integer.MAX_VALUE;
        
        queue.add(start);
        visited[start[0]][start[1]] = true;
        
        while(!queue.isEmpty()) {
            int[] now = queue.poll();
            
            if (now[0] == end[0] && now[1] == end[1]) {// end 도착
                count = Math.min(count, now[2]); // 최소 이동
                break;
            } 
            
            for(int i = 0; i < 4; i ++) {
                int[] next = find_pos(now, dx[i], dy[i], board);
                // System.out.println(Arrays.toString(next));
                if(!visited[next[0]][next[1]]){
                    visited[next[0]][next[1]] = true;
                    queue.add(next);
                }
            }
        }
        
        if(count == Integer.MAX_VALUE){
            count = -1;
        }
        return count;
        
    }
    
    public int[] find_pos(int[] now, int dx, int dy, String[] board){
        int[] next = new int[3];
        int x = now[0];
        int y = now[1];

        int nx = x;
        int ny = y;

        while (true) {
            int tx = nx + dx;
            int ty = ny + dy;

            if (tx < 0 || tx >= board.length || ty < 0 || ty >= board[0].length()) { // 벽
                break;
            }

            if (board[tx].charAt(ty) == 'D') { // 장애물
                break;
            }

            nx = tx;
            ny = ty;
        }

        next[0] = nx;              
        next[1] = ny;
        next[2] = now[2] + 1;

        return next;
    }   
}